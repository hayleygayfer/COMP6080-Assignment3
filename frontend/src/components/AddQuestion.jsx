import React from 'react';
import PropTypes from 'prop-types';
import '../App.css'
import API from '../api.js';
const api = new API('http://localhost:5005');

function AddQuestion (id) {
  const token = localStorage.getItem('token');
  const [gameData, setGameData] = React.useState('');
  const [questionString, setQuestionString] = React.useState('');
  const [questionAnswers, setQuestionAnswers] = React.useState('');
  const [timeLimit, setTimeLimit] = React.useState('');
  const [questionMedia, setQuestionMedia] = React.useState('');
  const gameId = id.input;

  const getGameQuestionsRequest = async (quizId) => {
    try {
      const request = await api.makeAPIRequest(`admin/quiz/${quizId}`, token, 'GET', '', '');
      if (request.status === 200) {
        console.log('Got Game Data');
        const data = await request.json();
        setGameData(data);
      } else throw request.status
    } catch (error) {
      alert(`Invalid Quiz Request: ${error}`);
      console.log(error);
    }
  }

  let questionEditor = [];
  const openQuestionEditor = () => {
    getGameQuestionsRequest(gameId);
    questionEditor.push(<>
      <div>
        Question: <input
          type="text"
          onChange={e => setQuestionString(e.target.value)}
          value={questionString}
        /><br/>
        Time Limit: <input
          type="text"
          onChange={e => setTimeLimit(e.target.value)}
          value={timeLimit}
        /><br/>
        Answers: <input
          type="text"
          onChange={e => setQuestionAnswers(e.target.value)}
          value={questionAnswers}
        /><br/>
        Image or Video: <input
          type="text"
          onChange={e => setQuestionMedia(e.target.value)}
          value={questionMedia}
        /><br/>
      </div>
      <button className='button' onClick={addGameQuestionsRequest(gameId)}> Submit Question </button>
    </>)
  }

  const newQuestion = {
    questionString: questionString,
    answers: questionAnswers,
    timeLimit: timeLimit,
    imgSource: questionMedia
  };

  const addGameQuestionsRequest = async (quizId) => {
    if (!gameData) return 1;
    gameData.questions.push(newQuestion);

    try {
      const request = await api.makeAPIRequest(`admin/quiz/${quizId}`, token, 'PUT', '', gameData);
      if (request.status === 200) {
        console.log('Updated Game Data');
        questionEditor = [];
      } else throw request.status
    } catch (error) {
      alert(`Invalid Update Question Request: ${error}`);
      console.log(error);
    }
  }

  return (<>
    <button className='button' onClick={openQuestionEditor()}> Open Editor </button>
    {questionEditor}
  </>);
}

AddQuestion.propTypes = {
  gameId: PropTypes.number
};

export default AddQuestion;
