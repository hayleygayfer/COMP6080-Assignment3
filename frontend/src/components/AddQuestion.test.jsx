import { shallow } from 'enzyme';
import * as React from 'react';
import { AddQuestion } from './AddQuestion';
import renderer from 'react-test-renderer';

describe('AddQuestion', () => {
  const mockCallBack = jest.fn();
  const addQInput = {
    game: "",
    id: 297083846
  }
  const dodgyInput = {
    game: "",
    id: 1234567
  }
  const correctHTML = '<div>Question: <input type="text" value=""/><br/>Time Limit: <input type="text" value=""/><br/><div>Answer: <input type="text" value=""/><br/><button class="button">Add Answer as Correct</button><button class="button">Add Answer as Incorrect</button><br/>All Answers:<br/><br/>Correct Answers:<br/><br/></div>Image or Video: <input type="text" value=""/><br/></div><button class="button"> Submit Question </button>'

  it('triggers onClick event handler when clicked', () => {
    const buttonData = shallow((<AddQuestion input={addQInput}/>)).first('button');
    const button = shallow((<buttonData onClick={mockCallBack}>Ok!</buttonData>));
    button.simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('returns specific html', () => {
    const wrapper = shallow((<AddQuestion input={addQInput}/>));
    expect(wrapper.html()).toEqual(correctHTML);
  });

  it('returns same inital output with dodgy game id', () => {
    const wrapper = shallow(<AddQuestion input={dodgyInput}/>);
    expect(wrapper.html()).toEqual(correctHTML);
  });

  // SNAPSHOT TESTING TO INCASE ANYTHING CHANGES:
  it('renders correctly when no click (snapshot)', () => {
    const wrapper = renderer.create(<AddQuestion input={addQInput}/>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
