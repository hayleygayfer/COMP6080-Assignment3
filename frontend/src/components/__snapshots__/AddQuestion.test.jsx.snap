// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddQuestion renders correctly when no click (snapshot) 1`] = `
Array [
  <div>
    Question: 
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    Time Limit: 
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <div>
      Answer: 
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
      <br />
      <button
        className="button"
        onClick={[Function]}
      >
        Add Answer as Correct
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        Add Answer as Incorrect
      </button>
      <br />
      All Answers:
      <br />
      <br />
      Correct Answers:
      <br />
      <br />
    </div>
    Image or Video: 
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
  </div>,
  <button
    className="button"
    onClick={[Function]}
  >
     Submit Question 
  </button>,
]
`;
